### PR TITLE
perf(dsa): tile indexer heads in the unfused score fallback

### PR DESCRIPTION
**Pull request by XiaomingFun233**
# What does this PR do?

  Bound head-tiled score accumulation in the DSA PyTorch fallback so the
  unfused indexer scorer launches one BMM per head tile instead of one BMM
  per head.

  :warning: Fallback-only change in the cuDNN wrapper. No fused-kernel or
  training-path change.

  ## Issue tracking

  Linked issue: Fixes #<issue-number>

  ## Summary

  `_compute_indexer_scores_chunk_with_global_rows` currently does:

  for h in range(num_heads):
  scores += ReLU(Q[h] @ K) * weight[h]   # one BMM per head

  Indexer heads share one key head. This PR folds up to 8 query heads into
  the BMM row dimension:

  Q tile [B, SQ, T, D] -> [B, SQT, D]
  K      [B, D, SK]                 # not replicated
  BMM -> [B, SQT, SK] -> [B, SQ, T, SK]
  ReLU, multiply matching weights, sum(dim=2), accumulate

  Tile size:

  bytes_per_head = B * SQ * SK * 4
  T = max(1, min(num_heads, 8, 256MiB // bytes_per_head))

  The 256 MiB budget covers only the temporary `[B, SQ, T, SK]` scores, not
  the output, the reduction, or Q/K/weight casts. A remainder tile smaller
  than `T` is handled on the last iteration. Each tile is `del`'d before the
  next GEMM. Q/K/weights and accumulation stay FP32.

  Fused cuDNN kernels are not touched.

  ## Behavioral notes

  - Algorithm is the same weighted ReLU score.
  - FP32 reduction order changes from per-head `add_` to tile `sum` then
    `add_`. Scores match within tolerance, not bitwise. Near-tied top-k
    entries can reorder.
  - When one head already exceeds 256 MiB, `T=1` and scores are bitwise
    identical to the previous loop. Explicit `del` can reduce peak because
    the old loop transiently kept the previous BMM output.

  ## Performance

  NVIDIA H200, PyTorch 2.13.0+cu130, CUDA 13.0, TF32 off, inference
  mode. `B=1, H=64, D=128`. Keys/positions cached; includes masking;
  excludes top-k/attention. Speedup vs the current per-head loop.

  BF16 `SQ x SK` speedup:

  | SQ \ SK | 256 | 1K | 4K | 8K | 16K | 32K | 64K |
  |---|---:|---:|---:|---:|---:|---:|---:|
  | 1 | 6.3x | 6.6x | 6.6x | 6.1x | 6.1x | 6.1x | 5.9x |
  | 128 | 6.0x | 6.0x | 5.8x | 3.3x | 1.7x | 1.2x | 1.2x |
  | 512 | 6.1x | 5.9x | 1.7x | 1.3x | 1.2x | 1.1x | 0.98x |
  | 1024 | 6.1x | 3.3x | 1.3x | 1.2x | 1.1x | 0.97x | 1.00x |

  - Clear win: `SQ*SK <= 4M` (about 1.5–6.6x). Decode `SQ=1` stays ~6x through `SK=64K`.
  - Small win: `4M < SQ*SK <= 16M` (about 1.1–1.5x).
  - Neutral / slight loss: `SQ*SK >= 32M`. Budget already selects `T=2` or `T=1`.
    `T=1` (`SQ=1024, SK=65536`) is bitwise identical; peak extra memory drops
    from ~768 MiB to ~512 MiB.

  This is expected: long rows are already compute-bound on a single-head
  `[B,SQ,D] x [B,D,SK]` BMM. Head tiling does not split `SK`; it only adds a
  `[B,SQ,T,SK]` ReLU / strided reduce. The 256 MiB cap is what stops the
  fallback from keeping that overhead on the largest shapes.

  Selected `T` examples: `[1,512,4096] -> 8`, `[2,512,16384] -> 4`,
  `[1,1024,65536] -> 1`.

  ## Test plan

  - [x] I have added relevant unit tests
  - [ ] I have added relevant functional tests
  - [x] I have added proper typing to my code
  - [ ] I have added relevant documentation
  - [ ] I have run `tools/autoformat.sh` on my PR

  Unit tests in
  `tests/unit_tests/transformer/experimental_attention_variant/test_dsa_indexer_head_tiles.py`:

  - CPU/CUDA, FP32/BF16
  - 1 / 7 / 17 heads (partial last tile)
  - budget caps 1 / 4 / 8
  - strided Q/K/W
  - cached and uncached `k_bdk` / `key_positions`
  - causal ratio/offset and explicit varlen
  - score parity (`rtol=atol=2e-5`) and top-k index membership
  - budget selector: `(1,512,4096)->8`, `(2,512,16384)->4`, `(1,8192,65536)->1`

  Locally:

  CUDA_VISIBLE_DEVICES=0 python -m torch.distributed.run --nnodes 1 --nproc-per-node 1
  -m pytest -q tests/unit_tests/transformer/experimental_attention_variant/test_dsa_indexer_head_tiles.py

  76 passed on H200 and on B200 ( PyTorch 2.11.0+cu130).
  Existing indexer top-k tests in `test_dsa_native_parity.py`: 22 passed, 1
  skipped (`flash_mla` unavailable). No fused-kernel or end-to-end training run.

### CODE REVIEW

**Comment by XiaomingFun233** [comment]
@asolergi-nv Hi, I found you on the current mcore oncall schedule.
  Could you please help authorize CI for this PR and route it to the appropriate reviewer? Local CUDA tests passed on H200 and B200. Thanks!

**Comment by XiaomingFun233** [comment]
@Connor-XY Could you please help authorize CI for this PR and route it to the appropriate reviewer?
